### ISSUE TITLE
Dynamically allocated output messages, arrays, and reference counting

### ISSUE THREAD

**Issue by edwardalee** [icyphy/lingua-franca#71]
TLDR: I'm proposing that we add to the Lingua Franca syntax the ability to specify a type of the form "array(type)", where "type" is any other type.  There is some sample code in test/C/Array*.lf that compiles and runs and shows what the code generator needs to do.

This will enable us to write a reactor that produces a dynamically allocated array as follows:
```
reactor Source {
    output out:array(int);
    reaction(startup) -> out {=
        // Dynamically allocate an output array of length 3.
        int* out_value = new_array(out, 3);
        // Above returns the array, which then must be populated.
        out_value[0] = 0;
        out_value[1] = 1;
        out_value[2] = 2;
    =}
}
```
The output produced will have a reference count and the memory will be automatically freed after the last use of the array by a downstream reactor.

Here is an example of a downstream reactor:
```
reactor Print {
    input in:array(int);
    reaction(in) {=        
        printf("Received: [");
        for (int i = 0; i < in_length; i++) {
            if (i > 0) printf(", ");
            printf("%d", in[i]);
        }
        printf("]\n");
    =}
}
```
Note that "in" is bound to the array and "in_length" to its length.

A reactor may want to *modify* an input array, in which case it should request a writable copy. This can be optimized so the writable copy is not a copy if the reference count is exactly one.  So, for example, we could write:
```
reactor Scale() {
    input in:array(int);
    output out:array(int);
    reaction(in) -> out {=
        int* modified = writable_copy(in);
        for(int i = 0; i < in_length; i++) {
            modified[i] = in_[i] * 2;
        }
        set_array(out, modified, in_length);
    =}
}
```
The writable_copy() function returns the original array if this reactor is the last one using the array.
Note that the set_array() needs to provide the length because C arrays do not include their length.

I have pushed an implementation of this that works without modification to the grammar or code generator, illustrating that it should be possible to make the above work properly. There is one big subtlety that bothers me, however:

When writable_copy() returns the original array, the user may get confused thinking that it is actually a copy.  Perhaps there is a better name for this function than writable_copy(). Also, in the current implementation, if there are several reactions reading the same input and the first one modifies the input, subsequent reactions will see the *modified* input, not the original. In my implementation, the reference count counts downstream *reactors* not downstream *reactions*. Perhaps it should count *reactions* and decrement the reference count after each reaction.

**Comment by lhstrh** [comment]
Some comments/questions:
 - Rather than `writable_copy` we could have a `get` that returns a pointer accordingly depending on whether the input port argument has a `mutable` or `writable` modifier. Note that with this approach (at the cost of performance overhead) we could actually enforce the policy. We could capture violations as early as at the end of a reaction.
- Would we complement this approach with an "object" type that would map to `struct`?
- How do we deal with more complex structures than arrays of which the elements are primitive types?
- The array type and how it is handled resembles the `repeat` fields in Protobufs a lot; this suggests that we might be better off making it compliant.

**Comment by edwardalee** [comment]
Good idea to have a mutable or writable modifier on an input declaration. This could also help guide scheduling, since it would be advantageous to postpone invocation of a reactor that has such a modifier until after other reactors that have no such modifier to avoid the copy.

Also, supporting something like an object type or struct along with array seems necessary to me. But I'm really hesitant to add complexity to the Lingua Franca syntax, particularly since this could end up being rather type specified.

How about this alternative: We allow a "type" in LF to have the following forms:
* id
* id*
* id[]
* in[number]
These forms need not be used by every target language. They are merely *allowed*. Any type designation is target language specific in any case.  Of course, we could specify these using the following syntax, which is already allowed, but it reads really badly:
* {=id*=}
The above looks a bit like a swear word.

The C generator would handle the last three type designations specially, using a token with reference counting for each.

**Comment by edwardalee** [comment]
I have implemented a 'mutable' modifier to an input specification and it now works with reference counting. Copying is avoided when an output has only one destination, but more work needs to be done to avoid copying when there is more than one destination. Specifically, we have to modify the precedence graph if we want to avoid copying. We would need to force the reaction that decrements the reference count to execute before reactions in parallel reactors that read the same input. The ArrayParallel.lf test illustrates this. It is unable to avoid copying.

The syntax looks pretty clean to me now. The following reactor takes an array in, scales it by a constant, and sends to the output:
```
reactor Scale(scale:int(2)) {
    mutable input in:int[];
    output out:int[];
    reaction(in) -> out {=
        for(int i = 0; i < in_length; i++) {
            in[i] = in[i] * self->scale;
        }
        set_array(out, in, in_length);
    =}
}
```
Note that the set_array() function needs to specify the array length because arrays in C do not carry their own length.

The array is dynamically allocated upstream, freed downstream, and copying is avoided if there are no parallel users of the array.

Next, we need to provide similar support for structs rather than arrays.

**Comment by edwardalee** [comment]
This is done, with support for structs and arrays.
